From 48ffcf64c5a7978ae9720d8c10e078fc0c9aa792 Mon Sep 17 00:00:00 2001
From: Vitalii Kulikov <solk@solk.org.ua>
Date: Mon, 16 Jan 2017 22:36:29 +0100
Subject: [PATCH 30/44] minui: Skip EV_REL input devices.

This is causing recovery to skip real input devices on some
samsung phones where sensors are registered as input devices.
So there more then 16 of them. (ex. ks01lte)
And EV_REL input devices already disabled in recovery ui.cpp if
BOARD_RECOVERY_NEEDS_REL_INPUT is not set. So do same here not to exceed
the limit of MAX_DEVICES

Change-Id: If3d6e29d00229278a8ef3dfa445393c9f3d5f361
(cherry picked from commit 331afccf4a9b81ddf22500872b7260b532e15c41)
(cherry picked from commit e7597e934e909112c462024e49a5f43af9bbe341)
---
 minui/Android.mk | 4 ++++
 minui/events.cpp | 9 +++++++--
 2 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/minui/Android.mk b/minui/Android.mk
index 72f46662..bc9357ad 100644
--- a/minui/Android.mk
+++ b/minui/Android.mk
@@ -74,6 +74,10 @@ ifneq ($(BOARD_RECOVERY_NEEDS_FBIOPAN_DISPLAY),)
   LOCAL_CFLAGS += -DBOARD_RECOVERY_NEEDS_FBIOPAN_DISPLAY
 endif
 
+ifneq ($(BOARD_RECOVERY_NEEDS_REL_INPUT),)
+  LOCAL_CFLAGS += -DBOARD_RECOVERY_NEEDS_REL_INPUT
+endif
+
 include $(BUILD_STATIC_LIBRARY)
 
 # libminui (shared library)
diff --git a/minui/events.cpp b/minui/events.cpp
index 375ea5ca..e02aa2e0 100644
--- a/minui/events.cpp
+++ b/minui/events.cpp
@@ -78,9 +78,14 @@ int ev_init(ev_callback input_cb, bool allow_touch_inputs) {
         continue;
       }
 
-      // We assume that only EV_KEY, EV_REL, and EV_SW event types are ever needed. EV_ABS is also
+      // We assume that only EV_KEY, and EV_SW event types are ever needed. EV_ABS is also
       // allowed if allow_touch_inputs is set.
-      if (!test_bit(EV_KEY, ev_bits) && !test_bit(EV_REL, ev_bits) && !test_bit(EV_SW, ev_bits)) {
+      // EV_REL should be enabled explicitly in device tree.
+      if (!test_bit(EV_KEY, ev_bits) &&
+#ifdef BOARD_RECOVERY_NEEDS_REL_INPUT
+          !test_bit(EV_REL, ev_bits) &&
+#endif
+          !test_bit(EV_SW, ev_bits)) {
         if (!allow_touch_inputs || !test_bit(EV_ABS, ev_bits)) {
           close(fd);
           continue;
-- 
2.17.1

